subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java

```diff
@@ -102,68 +106,76 @@ private void deleteDistributions(Collection<File> dirs) {
     }
 
     private Multimap<GradleVersion, File> determineChecksumDirsByVersion() {
-        Multimap<GradleVersion, File> result = ArrayListMultimap.create();
+        Multimap<GradleVersion, File> checksumDirsByVersion = ArrayListMultimap.create();
         for (File dir : listDirs(distsDir)) {
             for (File checksumDir : listDirs(dir)) {
-                try {
-                    GradleVersion gradleVersion = determineGradleVersionFromBuildReceipt(checksumDir);
-                    result.put(gradleVersion, checksumDir);
-                } catch (Exception e) {
-                    // TODO see https://github.com/gradle/gradle-private/issues/1379
-                    // LOGGER.debug("Could not determine Gradle version for {}", checksumDir, e);
+                Result<GradleVersion> result = determineGradleVersionFromBuildReceipt(checksumDir);
+                if (result.hasValue()) {
+                    checksumDirsByVersion.put(result.getValue(), checksumDir);
+                } else {
+                    LOGGER.debug("Could not determine Gradle version for {}: {}", checksumDir, result.getError());
```

Comment:
I don't understand this change. To me it looks like the logic is doing exactly the same thing as before. Just using a monad instead of Exception, but still logging everything. Can you point me to the difference?

Reply:
Okay, I see the stacktrace now. I don't understand the misusing exceptions point though.

Reply:
What I mean by "misuse" is throwing an exception, e.g. from

https://github.com/gradle/gradle/blob/40ad77c5cd530a3536ce9fc34c23bd64657746f0/subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java#L140

only to catch and log it two methods further up:

https://github.com/gradle/gradle/blob/40ad77c5cd530a3536ce9fc34c23bd64657746f0/subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java#L111-L113

If I understand you correctly, you're not a big fan of the `Result` class I added, are you?

Reply:
If we were in a more functional language, the monad would be the idiomatic choice. But for Java I think the exception is just fine. You are handling a failure case after all and if the code throwing the exception was in another class, no one would even raise an eyebrow I think.